contracts/rewards/SubgraphAvailabilityManager.sol

```diff
@@ -0,0 +1,206 @@
+
+    // -- State --
+
+    uint256 public voteTimeLimit;
+    address[] public oracles;
+    mapping(bytes32 => mapping(uint256 => uint256)) public lastDenyVote;
+    mapping(bytes32 => mapping(uint256 => uint256)) public lastAllowVote;
+
+    // -- Events --
+
+    /**
+     * @dev Emitted when an oracle is set
+     */
+    event OracleSet(uint256 indexed index, address indexed oracle);
+
+    /**
+     * @dev Emitted when the vote time limit is set
+     */
+    event VoteTimeLimitSet(uint256 voteTimeLimit);
+
+    /**
+     * @dev Emitted when an oracle votes to deny or allow a subgraph
+     */
+    event OracleVote(
+        bytes32 indexed subgraphDeploymentID,
+        bool deny,
+        uint256 oracleIndex,
+        uint256 timestamp
+    );
+
+    // -- Modifiers --
+
+    modifier onlyOracle(uint256 _oracleIndex) {
+        require(_oracleIndex < maxOracles, "SAM: index out of bounds");
+        require(msg.sender == oracles[_oracleIndex], "SAM: caller must be oracle");
+        _;
+    }
+
+    // -- Constructor --
+
+    /**
+     * @dev Contract constructor
+     * @param _governor Account that can set or remove oracles and set the vote time limit
+     * @param _rewardsManager Address of the RewardsManager contract
+     * @param _maxOracles Maximum number of oracles
+     * @param _executionThreshold Number of votes required to execute a deny or allow call to the RewardsManager
+     * @param _voteTimeLimit Vote time limit in seconds
+     */
+    constructor(
+        address _governor,
+        address _rewardsManager,
+        uint256 _maxOracles,
+        uint256 _executionThreshold,
+        uint256 _voteTimeLimit
+    ) {
+        require(_governor != address(0), "SAM: governor must be set");
+        require(_rewardsManager != address(0), "SAM: rewardsManager must be set");
+        require(_maxOracles > 1, "SAM: maxOracles must be greater than 1");
+        require(_executionThreshold > 1, "SAM: executionThreshold must be greater than 1");
+
+        Governed._initialize(_governor);
+        rewardsManager = IRewardsManager(_rewardsManager);
+
+        maxOracles = _maxOracles;
+        executionThreshold = _executionThreshold;
+        voteTimeLimit = _voteTimeLimit;
+        oracles = new address[](_maxOracles);
+
+        // Initialize oracles array with empty addresses
+        for (uint256 i = 0; i < _maxOracles; i++) {
+            oracles[i] = address(0);
+        }
+    }
+
+    // -- Functions --
+
+    /**
+     * @dev Set the vote time limit
+     * @param _voteTimeLimit Vote time limit in seconds
+     */
+    function setVoteTimeLimit(uint256 _voteTimeLimit) external onlyGovernor {
+        voteTimeLimit = _voteTimeLimit;
+        emit VoteTimeLimitSet(_voteTimeLimit);
+    }
+
+    /**
+     * @dev Set oracle address with index
+     * @param _index Index of the oracle
+     * @param _oracle Address of the oracle
+     */
+    function setOracle(uint256 _index, address _oracle) external onlyGovernor {
```

Comment:
Great point. I like the idea of cleaning up the oracles votes after replacing them but to do this we'll have to keep track of all subgraph deployment IDs and we'll have to iterate over that array and clean up one by one unless I'm missing an easier way of implementing this.

For `setVoteTimeLimit` I understand that with changing it a vote that passed maybe shouldn't have or the opposite but not sure if/how we can do something about that. 

In any case I my knowledge of the oracle is that it runs every X interval and checks current state of all subgraphs and sets/votes if it's different so it would correct itself at some point. 

What do you think?

Reply:
True, wiping the votes is not easy! 

Assuming that the oracle eventually corrects itself, and assuming a short enough time limit window (i.e `voteTimeLimit` << 1 epoch) I think we can probably "ignore" this problems, and accept that there is a very low probability that some rewards get misplaced. If we do this we should document it because auditors will likely bring it up.


Tagging @pcarranzav for his thoughts.

Reply:
I think the fact that there is an expiration window makes this less important - we can accept that the last votes from an oracle before it's replaced will be valid until their expiration.
Oracles should always try to reconcile the state in RewardsManager to what they think should be denied, so they should keep retrying constantly if something doesn't match.

That being said, one easy way to make all votes expire is to add a nonce that can be incremented, so instead of
```
lastDenyVote[subgraphDeploymentId][oracleIndex] = timestamp;
```

we can have
```
lastDenyVote[currentNonce][subgraphDeploymentId][oracleIndex] = timestamp;
```

then whenever any oracles are set we can increment currentNonce and start a fresh batch of voting.